db/commitlog/commitlog.cc

```diff
@@ -1741,41 +1741,64 @@ future<> db::commitlog::segment_manager::delete_file(const sstring& filename) {
+                    co_await ext->before_delete(filename);
+                }
+            }
 
-    return parallel_for_each(i, e, [this](auto& filename) {
-        auto f = make_ready_future();
-        auto exts = cfg.extensions;
-        if (exts && !exts->commitlog_file_extensions().empty()) {
-            f = parallel_for_each(exts->commitlog_file_extensions(), [&](auto& ext) {
-                return ext->before_delete(filename);
-            });
-        }
-        return f.finally([&] {
             // We allow reuse of the segment if the current disk size is less than shard max.
-            auto usage = totals.total_size_on_disk;
-            if (!_shutdown && cfg.reuse_segments && usage <= max_disk_size) {
-                descriptor d(next_id(), "Recycled-" + cfg.fname_prefix);
-                auto dst = this->filename(d);
-
-                clogger.debug("Recycling segment file {}", filename);
-                // must rename the file since we must ensure the
-                // data is not replayed. Changing the name will
-                // cause header ID to be invalid in the file -> ignored
-                return rename_file(filename, dst).then([this, dst]() mutable {
-                    auto b = _recycled_segments.push(std::move(dst));
-                    assert(b); // we set this to max_size_t so...
-                    return make_ready_future<>();
-                }).handle_exception([this, filename](auto&&) {
-                    return delete_file(filename);
-                });
+            if (!_shutdown && cfg.reuse_segments) {
+                auto usage = totals.total_size_on_disk;
+                auto recycle = usage <= max_disk_size;
+
+                // if total size is not a multiple of segment size, we need
+                // to check if we are the overlap segment, and noone else
+                // can be recycled. If so, let this one live so allocation
+                // can proceed. We assume/hope a future delete will kill
+                // files down to under the threshold, but we should expect
+                // to stomp around nearest multiple of segment size, not 
+                // the actual limit.
+                if (!recycle && _recycled_segments.empty() && files.empty()) {
+                    auto size = co_await seastar::file_size(filename);
+                    recycle = (usage - size) <= max_disk_size;
+                }
+
+                if (recycle) {
+                    descriptor d(next_id(), "Recycled-" + cfg.fname_prefix);
+                    auto dst = this->filename(d);
+
+                    clogger.debug("Recycling segment file {}", filename);
+                    // must rename the file since we must ensure the
+                    // data is not replayed. Changing the name will
+                    // cause header ID to be invalid in the file -> ignored
+                    try {
+                        co_await rename_file(filename, dst);
```

Comment:
Not sure how probable this is, but what if renaming of _every_ segment from `files` fails? No segments will be sent to recycling and we will get stuck just like before?

Reply:
It is a problem. The allocate waiter should wait for either of delete/recycle. But I could not find a `wait-for-either` primitive any longer, so not part of this.

Reply:
I'm OK with fixing it in another PR, but I think we should do it soon. Please create an issue/send a followup so that we don't lose track of the problem.

Reply:
I added both a second issue, and a fix patch to this PR. Two for one. Please re-review.

Reply:
Usually we wrap file operations with commit_error_handler which shuts down operations. Why is this not done here?

@piodul if things like renaming files stop working, the node is cooked.

Reply:
The code here was never error handled before. Probably an oversight. Though I guess we could add it now.

Reply:
Speak if you want me to add wrapping in this patch set. But it seems slightly unrelated.

Reply:
I think wrapping the rename operation in commit_error_handler falls under "handle commitlog recycle errors", and this PR is already marked as a fix for that issue. Therefore I think it is in the current scope of this PR.

Reply:
Actually, I just realize that both rename_file and delete_file both use the error handler wrapping. So it is already handled.

Reply:
Oh, I see - rename_file is a method of the segment_manager, which is named in the same way as the seastar function.